cpp/src/hash/hashing.cu

```diff
@@ -213,19 +226,196 @@ void compute_row_output_locations(size_type * __restrict__ row_partition_numbers
+    }
+  }
+
+  // Fetch the offset in the output buffer of each partition in this thread block
+  for (size_type ipartition = threadIdx.x; ipartition < num_partitions;
+       ipartition += blockDim.x) {
+    partition_offset_global[ipartition]
+      = scanned_block_partition_sizes[ipartition * gridDim.x + blockIdx.x];
+  }
+
+  __syncthreads();
+
+  // Fetch the input data to shared memory
+  for (size_type row_number = threadIdx.x + blockIdx.x * blockDim.x;
+       row_number < num_rows; row_number += blockDim.x * gridDim.x) {
+    size_type const ipartition = row_partition_numbers[row_number];
+
+    block_output[
+      partition_offset_shared[ipartition] + row_partition_offset[row_number]
+    ] = input_iter[row_number];
+  }
+
+  __syncthreads();
+
+  // Copy data from shared memory to output using 16 threads for each partition
+  constexpr int nthreads_partition = 16;
+  static_assert(OPTIMIZED_BLOCK_SIZE % nthreads_partition == 0,
+    "BLOCK_SIZE must be divisible by number of threads");
+  
+  for (size_type ipartition = threadIdx.x / nthreads_partition; ipartition < num_partitions;
+       ipartition += OPTIMIZED_BLOCK_SIZE / nthreads_partition) {
+    size_type const nelements_partition = partition_offset_shared[ipartition + 1]
+      - partition_offset_shared[ipartition];
+
+    for (size_type row_offset = threadIdx.x % nthreads_partition;
+         row_offset < nelements_partition; row_offset += nthreads_partition) {
+      output_buf[partition_offset_global[ipartition] + row_offset]
+        = block_output[partition_offset_shared[ipartition] + row_offset];
+    }
+  }
+}
+
+rmm::device_vector<size_type> compute_gather_map(
+    size_type num_rows,
+    size_type num_partitions,
+    size_type const * row_partition_numbers,
+    size_type const * row_partition_offset,
+    size_type const * block_partition_sizes,
+    size_type const * scanned_block_partition_sizes,
+    size_type grid_size,
+    cudaStream_t stream)
+{
+  // We need 3 chunks of shared memory:
+  // 1. BLOCK_SIZE * ROWS_PER_THREAD elements of size_type for copying to output
+  // 2. num_partitions + 1 elements of size_type for per-block partition offsets
+  // 3. num_partitions + 1 elements of size_type for global partition offsets
+  int const smem = OPTIMIZED_BLOCK_SIZE * OPTIMIZED_ROWS_PER_THREAD * sizeof(size_type)
+    + (num_partitions + 1) * sizeof(size_type) * 2;
+
+  rmm::device_vector<size_type> gather_map(num_rows);
+  auto iter = thrust::make_counting_iterator(0);
+
+  copy_block_partitions
+    <<<grid_size, OPTIMIZED_BLOCK_SIZE, smem, stream>>>(
+      iter, gather_map.data().get(), num_rows,
+      num_partitions, row_partition_numbers, row_partition_offset,
+      block_partition_sizes, scanned_block_partition_sizes
+  );
+
+  return gather_map;
+}
+
+struct copy_block_partitions_dispatcher{
+  template <typename DataType,
+      std::enable_if_t<is_fixed_width<DataType>()>* = nullptr>
+  std::unique_ptr<column> operator()(
```

Comment:
The body of this operator and `compute_gather_map` are very similar. Can you factor anything out?

Reply:
Ok, done